### PR TITLE
Orphan blocks 

### PR DESCRIPTION
**Pull request by eesmerdag**

### CODE REVIEW

**Comment by eesmerdag** [comment]
https://www.notion.so/figmentnetworks/Mina-same-canonical-block-shows-2-producers-b6f619c3c666465fb6bb24d4708fc1a5

**Review by sosedoff — COMMENTED** [review]
Changes look good. I think we'll need to address the following things too:

- Transaction search needs to return results from canonical transactions unless we want to expose a `canonical` filter on that endpoint. 
- Network stats queries needs to be adjusted to query canonical blocks only. 

^ Items above are related to Hubble to we need to figure out if any of the changes are going to break anything.

File: store/blocks.go [hunk 1]
```diff
@@ -30,9 +30,14 @@ func (s BlocksStore) FindByHash(hash string) (*model.Block, error) {
...
 }
 
-// FindByHeight returns a block with the matching height
+// FindByHeight returns a canonical block with the matching height
 func (s BlocksStore) FindByHeight(height uint64) (*model.Block, error) {
-	return s.FindBy("height", height)
+	result := model.Block{}
+
+	scope := s.db.Limit(1)
+	scope = scope.Where("height = ?", height).Where("canonical = ?", true)
```
[the remark this answers is not in the document]
Comment (eesmerdag) [reply store/blocks.go:38]: 👍

File: store/blocks.go [hunk 2]
```diff
@@ -74,3 +79,13 @@ func (s BlocksStore) AvgTimes(limit int64) ([]byte, error) {
 func (s BlocksStore) Stats(period uint, interval string) ([]byte, error) {
 	return jsonquery.MustArray(s.db, queries.BlocksStats, period, interval)
 }
+
+// UpdateCanonicalBlocks updates all blocks as non canonical at a height
+func (s BlocksStore) UpdateCanonicalBlocksAllFalseByHeight(height uint64) error {
+	return s.db.Exec(queries.UpdateCanonicalBlocksAllFalse, height).Error
```
[the remark this answers is not in the document]
Comment (eesmerdag) [reply store/blocks.go:85]: 👍

File: store/blocks.go [hunk 3]
```diff
@@ -74,3 +79,13 @@ func (s BlocksStore) AvgTimes(limit int64) ([]byte, error) {
...
+}
+
+// UpdateCanonicalBlock updates canonical at a height
+func (s BlocksStore) UpdateCanonicalBlock(hash string) error {
+	return s.db.Exec(queries.UpdateCanonicalBlock, hash).Error
```
[the remark this answers is not in the document]
Comment (eesmerdag) [reply store/blocks.go:90]: 👍

**Comment by eesmerdag** [comment]
> Changes look good. I think we'll need to address the following things too:
> 
> * Transaction search needs to return results from canonical transactions unless we want to expose a `canonical` filter on that endpoint.
> * Network stats queries needs to be adjusted to query canonical blocks only.
> 
> ^ Items above are related to Hubble to we need to figure out if any of the changes are going to break anything.

thanks for your comment. these are implemented in this PR

**Comment by pacmessica** [comment]
is this ready for merge? need @sosedoff to approve?

**Comment by sosedoff** [comment]
Couple of things i found after some testing:

- `/block` endpoint does not return the canonical block as it use do
- `/block/:height` returns 500 server error if block for a given height is not a canonical, fetching by hash still works
- Transactions search model has a field `Canonical` vs lowercase `canonical`

**Pull request merged**